LCB: Rename standard library types.

peter-b opened the pull request:
The current standard library uses lower-case type names.  This has
some unfortunate interactions with the current syntax implementation;
for example, the use of an "index" type means that it's not possible
to add any syntax that uses "index" as a keyword.

Initial draft of spec for new naming convention for standard library
types.

[Rendered version](https://github.com/peter-b/livecode/blob/feature-lcbtyperename/docs/specs/lcb-type-naming.md)

peter-b committed:
LCB: Rename standard library types.

The current standard library uses lower-case type names.  This has
some unfortunate interactions with the current syntax implementation;
for example, the use of an "index" type means that it's not possible
to add any syntax that uses "index" as a keyword.

Initial draft of spec for new naming convention for standard library
types.

peter-b commented:
Please bikeshed this as quickly as possible, because I need to start implementing it this afternoon if possible.

runrevmark commented:
Purple or Orange? ;)

runrevmark commented:
UIntPtr / IntPtr? (Potentially consider Pointer -> Ptr ?)

Also, we should potentially have universal names for the binary-fp types bearing in mind we'll also have decimal arithmetic at some point - BinFlt32 / BinFlt64? (Which would allow DecFlt32 / DecFlt64 etc.)

peter-b commented:
@runrevmark For now, I was aiming to simply provide names for stuff that we already have (in fact, I think the only exception to that rule is the fixed-width integer types that I'm not planning on actually implementing in this branch).

Can we consider additional native types at a later date / as required?

runrevmark commented:
I think it would be useful to put other potential ones in the spec - for discussion and so we can see the scope of the naming policy (the more examples we have the easier it is to see whether we have the 'right' - for some definition of 'correct' - idea or not). (Just mark the ones which aren't implemented as 'TODO').

The reason I brought up the BinFlt32 suggestion was that we already have double/float implemented so should probably consider universal aliases. After all, LCIndex / CInt / LCInteger will all be aliases for the current 'int' foreign type.

peter-b commented:
On the subject of `Pointer` -> `Ptr` rename: it seems to me that the aim of LiveCode is descriptiveness rather than conciseness; I think if I wanted to write very dense native code I'd write in C or Rust and then call it using Builder's FFI.

peter-b commented:
Currently amending the specification to add other types as suggested.

runrevmark commented:
@peter-b: I agree with descriptiveness - I'm sure there's a rule here somewhere...

IntPtr seems better somehow than IntPointer

BinFlt32 seems better somehow than BinaryFloat32

Perhaps Pointer can just be an exception :)

runrevmark commented:
Btw, you never said whether you preferred purple or orange ;)

peter-b committed:
LCB: Second draft of type naming specification.

Add more types, some of which we're not going to implement straight
away.

peter-b commented:
@runrevmark: Orange all the way :bicyclist: :bicyclist: :bicyclist: 

Maybe we add the "verbose" versions to start with and then we can add abbreviated synonyms if we decide they're too unwieldy?

I've pushed a patch which adds some (verbose!) additional types to the spec.

runrevmark commented:
Further to discussion offline - I've been disuaded from BinFloat... @peter-b made the point that hardware tends towards binary representations and Int32 implicitly implicitly implies a binary-encoded integer... Thus Float32 should also implicitly imply binary.

I also retract DecFloat in favour of Float32Dec - the latter is more consistent with the ZString naming.

runrevmark commented:
I'm happy with this to be applied to the existing types we define :)

peter-b committed:
LCB: Use "Pointer" instead of "pointer".

peter-b committed:
LCB: Use "LCUInteger" or "UInt32" instead of "uint", as appropriate.

peter-b committed:
LCB: Use "CInt"/"LCInteger"/"LCIndex"/"Int32" instead of "int"

peter-b committed:
LCB: Use "CFloat" instead of "float".

peter-b committed:
LCB: Use "CDouble" instead of "double".

peter-b committed:
LCB: Use "LCIndex"/"LCUIndex" instead of "index".

peter-b committed:
com.livecode.char: Correct signedness of offset before/after index

peter-b committed:
LCB: Use "LCUIndex" instead of "index".

peter-b committed:
LCB: Use "CBool" instead of "bool".

peter-b committed:
LCB: Use "ZStringNative" instead of "NativeCString".

peter-b committed:
com.livecode.foreign: New names for foreign types.

peter-b committed:
lc-compile: Allow "CBool" instead of "bool" as iterator function result

peter-b committed:
Merge branch 'develop' into feature-lcbtyperename

This required some changes to the new codeunit module in order to
complete the merge, including removing a bogus function and
updating a lot of handler definitions with the new type names.

Conflicts:
	libscript/src/foreign.mlc

peter-b commented:
@runrevmark: This branch now renames all existing uses of the old native types that I was able to find to use the new names, while keeping the deprecated old names around.

The main question now is to which module(s) to add the mixed-case versions of the type names (i.e. `string` → `String`).

runrevmark commented:
Having seen this I wonder if we should (slightly breaking convention) make LCInteger, LCInt... The rationale is the potential confusion between 'high-level' Integer (LCIntegerRef) type and a POD 'Int' type. Also we are proposing Int32 and CInt - and probably Int (for platform-sized integer?).

In regards to the aliases - they should go in the module for the type... So string -> String in com.livecode.string.

peter-b committed:
lc-compile: Prefer TitleCase names for core types.

Make the compiler recognise "String", "List", "Array", etc. as names
for language core types.  Continue to support the original lowercase
variants of the type names, but emit a deprecation warning for each
use.

Minor fix-ups to enable the test-suite to continue to pass (generally
variables or parameters named e.g. "String").

peter-b committed:
LCB: Use "Boolean" instead of "boolean"

peter-b committed:
LCB: Use "String" instead of "string".

peter-b committed:
LCB: Use "Number" instead of "number".

peter-b committed:
LCB: Use "List" instead of "list".

peter-b committed:
LCB: Use "Array" instead of "array".

peter-b committed:
LCB: Use "Integer" instead of "integer".

peter-b committed:
LCB: Use "Real" instead of "real".

peter-b committed:
LCB: Use "Data" instead of "data".

peter-b committed:
LCB: Rename "LCInteger" to "LCInt" (etc.)

Rename for better consistency with CInt/CUint.

peter-b committed:
tests: Update "lc-test" test file to use new type names.

peter-b commented:
@runrevmark I've modified the compiler to use titlecase core types (e.g. `List`); lowercase core types (e.g. `list`) are still supported, but emit a deprecation warning.  I've update all of the standard library and test suite accordingly.  Because the existing widgets should still all work I haven't touched the `extensions/` hierarchy at all.  Finally, I've renamed `LCInteger` to `LCInt` and `LCUInteger` to `LCUInt`.

runrevmark commented on docs/specs/lcb-type-naming.md:
```diff
@@ -0,0 +1,111 @@
+
+* `LCIndex`, `LCUIndex`: Indices (e.g. into lists or strings)
+
+* `LCInteger`, `LCUInteger`: Integers
```
This should probably be LCInt / LCUInt now :)

peter-b committed:
LCB: Rename "LCInteger" to "LCInt" in naming specification.

runrevmark commented:
Reviewed-by: @runrevmark.
